docs/proposals/api_design.md

```diff
@@ -0,0 +1,62 @@
+---
+title: "HTTP API Design"
+linkTitle: "HTTP API Design"
+weight: 1
+slug: http-api-design
+---
+
+# HTTP API Design
+
+## Overview
+
+The purpose of this design document is to propose a set of standards that should be the basis of the Cortex HTTP API. This document will outline the current state of the Cortex http api and describe limitations that result from the current approach. It will also outline a set of paradigms on how http routes should be created within Cortex.
+
+## Current Design
+
+As things currently stand, the majority of HTTP API calls exist under the `/api/prom` path prefix. This prefix is configurable. However, since this prefix is shared between all the modules which leads to conflicts if the Alertmanager is attempted to be run as as part of the single binary.
+
+## Proposed Design
+
+### Module-Based Routing
+
+Cortex incorporates three separate APIs: Alertmanager, Prometheus, and Cortex. Each of these APIs should use a separate route prefix that accurately describes the API. Currently, all of the api calls in Cortex reside under the configured http prefix. Instead the following routing tree is proposed:
+
+#### `/prometheus/*`
+
+Under this path prefix, Cortex will act as a Prometheus web server. It will host all of the required Prometheus api endpoints. For example to query cortex the endpoint `/prometheus/api/v1/query_range` will be used.
+
+#### `/alertmanager/*`
+
+Under this path prefix, Cortex will act as a Alertmanager web server. In this case, it will forward requests to the alertmanager and support the alertmanager API
+
+#### `/api/v1/*` -- The cortex API will exist under this path prefix.
+
+- `/push` & `/read` -- The Prometheus remote write endpoints will be fulfilled directly under the `/api/v1/*` path
+- `/user_stats` -- The per user stats endpoint will exist at this endpoint
+- `/chunks` 
+- `/rules/*`
+
+#### `/` -- Endpoints directly exposed under the root path. This is identical to current behaviour.
+
+- `/config`
+- `/ring`
+- `/ruler_ring`
+- `/compactor_ring`
+- `/ha-tracker`
+- `/all_user_stats`
+- `/flush`
+- `/shutdown`
+
+### Path Versioning
+
+Cortex will utilize path based versioning similar to both Prometheus and Alertmanager. This will allow future versions of the API to be released with changes over time. 
```

Comment:
Will `v1` be introduced at this stage or later? If at this stage, would be great listing the new versioned endpoints.

Reply:
`v1` will be introduced at this stage. I'll add a list of what it will look like.